File: UDP/udpserver.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <sys/wait.h>

#define PORT 6666
#define BUFFER_SIZE 1024
#define MAX_CLIENTS 10

int main() {
    char buffer[BUFFER_SIZE];
    struct sockaddr_in servaddr, cliaddr;
    int sockfd, len, num_clients = 0;
    printf("Escuchando en el puerto número: %d\n", PORT);
    
    // Crear socket UDP
    sockfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (sockfd == -1) {
        perror("Fallo en la creación del socket");
        exit(EXIT_FAILURE);
    }
    
    // Configurar dirección del servidor
    memset(&servaddr, 0, sizeof(servaddr));
    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    servaddr.sin_port = htons(PORT);
    
    // Enlazar dirección del servidor al socket
    if (bind(sockfd, (struct sockaddr *)&servaddr, sizeof(servaddr)) < 0) {
        perror("Fallo en el enlace del socket");
        exit(EXIT_FAILURE);
    }
    
    // Obtener e imprimir la dirección IP en la que está escuchando
    char ip_str[INET_ADDRSTRLEN];
    inet_ntop(AF_INET, &(servaddr.sin_addr), ip_str, INET_ADDRSTRLEN);
    printf("Escuchando en la IP: %s\n", ip_str);
    
    while (1) {
        // Recibir nombre del archivo del cliente
        len = sizeof(cliaddr);
        int n = recvfrom(sockfd, buffer, BUFFER_SIZE, 0, (struct sockaddr *)&cliaddr, (socklen_t *)&len);
        if (n < 0) {
            perror("recvfrom failed");
            exit(EXIT_FAILURE);
        }
        buffer[n] = '\0';
        printf("Recibiendo del cliente: %s\n", buffer);
        
        // Crear un proceso hijo para atender al cliente
        pid_t pid = fork();
        if (pid < 0) {
            perror("fork failed");
            exit(EXIT_FAILURE);
        } else if (pid == 0) {
            // Código del proceso hijo
            FILE *file = fopen(buffer, "r");
            if (file == NULL) {
                perror("Error al abrir el archivo");
                exit(EXIT_FAILURE);
            }
            
            char bufferArc[BUFFER_SIZE];
            while (fgets(bufferArc, BUFFER_SIZE, file) != NULL) {
                sendto(sockfd, bufferArc, strlen(bufferArc), 0, (struct sockaddr *)&cliaddr, sizeof(struct sockaddr_in));
            }
            sendto(sockfd, "FIN", 3, 0, (struct sockaddr *)&cliaddr, sizeof(struct sockaddr_in));
            printf("Archivo enviado al cliente.\n");
            fclose(file);
            exit(0); // Salir del proceso hijo
        } else {
            // Código del proceso padre
            num_clients++;
            if (num_clients >= MAX_CLIENTS) {
                // Esperar a que un proceso hijo termine
                int status;
                waitpid(-1, &status, 0);
                num_clients--;
            }
        }
    }
    
    close(sockfd);
    return 0;
}

File: UDP/udpserver.rb
require 'socket'
require 'thread'

PORT = 6666
BUFFER_SIZE = 1024
MAX_CLIENTS = 15

def handle_client(client_info)
  cliaddr = client_info[:cliaddr]
  filename = client_info[:filename]

  sockfd = Socket.new(:AF_INET, :SOCK_DGRAM, 0)

  file = File.open(filename, 'r') # Abrir el archivo en modo de lectura de texto
  if file.nil?
    puts "Error al abrir el archivo"
    return
  end

  # Enviar datos del archivo al cliente en bloques
  buffer = ''
  while (line = file.gets)
    sockfd.send(line, 0, cliaddr)
    print"#{line}\n"
  end
  sockfd.send("FIN", 0, cliaddr)
  puts "Archivo enviado al cliente."

  # Cerrar el archivo y el socket
  file.close
  sockfd.close
rescue => e
  puts "Error: #{e.message}"
end

def main
  buffer = ''
  threads = []
  client_count = 0

  puts "Escuchando en el puerto número: #{PORT}"

  # Crear socket UDP
  sockfd = Socket.new(:AF_INET, :SOCK_DGRAM, 0)

  # Configurar dirección del servidor
  servaddr = Socket.pack_sockaddr_in(PORT, '0.0.0.0')

  # Enlazar dirección del servidor al socket
  sockfd.bind(servaddr)

  # Obtener e imprimir la dirección IP en la que está escuchando
  ip_str = sockfd.local_address.ip_address
  puts "Escuchando en la IP: #{ip_str}"

  loop do
    # Recibir nombre del archivo del cliente
    buffer, cliaddr = sockfd.recvfrom(BUFFER_SIZE)
    puts "Recibiendo del cliente: #{buffer}"

    # Crear un nuevo hilo para manejar la solicitud del cliente
    if client_count < MAX_CLIENTS
      client_info = { cliaddr: cliaddr, filename: buffer.strip }
      thread = Thread.new { handle_client(client_info) }
      threads << thread
      client_count += 1
    else
      puts "Número máximo de clientes alcanzado. Solicitud rechazada."
    end
  end

  # Esperar a que todos los hilos terminen
  threads.each(&:join)

  sockfd.close
rescue => e
  puts "Error: #{e.message}"
end

main

